src/gtil/predict.cc

```diff
@@ -49,10 +50,16 @@ inline int NextNodeCategorical(float fvalue, const std::vector<uint32_t>& matchi
   if (std::isnan(fvalue)) {
     return default_child;
   }
-  const auto category_value = static_cast<uint32_t>(fvalue);
-  const bool is_matching_category = (
-      std::find(matching_categories.begin(), matching_categories.end(), category_value)
-      != matching_categories.end());
+  bool is_matching_category;
+  float max_representable_int = static_cast<float>(uint32_t(1) << FLT_MANT_DIG);
+  if (fvalue < 0 || std::fabs(fvalue) > max_representable_int || std::isinf(fvalue)) {
```

Comment:
> For NAN, one could swap the branches and invert the condition?

Not so easy, since result for NAN depends on the `default_left` flag. It's easier to explicitly check for NAN.